joinservice/internal/certcache/certcache.go

```diff
@@ -0,0 +1,165 @@
+
+	return &Client{
+		attVariant: attVariant,
+		log:        log,
+		kubeClient: kubeClient,
+		kdsClient:  kdsClient,
+	}
+}
+
+// CreateCertChainCache creates a certificate chain cache for the given attestation variant
+// and returns the cached certificates, if applicable.
+// If the certificate chain cache already exists, nothing is done.
+func (c *Client) CreateCertChainCache(ctx context.Context) (*CachedCerts, error) {
+	switch *c.attVariant {
+	case variant.AzureSEVSNP{}:
+		c.log.Debugf("Creating azure SEV-SNP certificate chain cache")
+		ask, ark, err := c.createCertChainCache(ctx, abi.VcekReportSigner)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create azure SEV-SNP certificate chain cache: %w", err)
+		}
+		return &CachedCerts{
+			ask: ask,
+			ark: ark,
+		}, nil
+	// TODO(derpsteb): Add AWS
+	default:
+		c.log.Debugf("No certificate chain caching possible for attestation variant %s", *c.attVariant)
+		return nil, nil
+	}
+}
+
+// CachedCerts contains the cached certificates.
+type CachedCerts struct {
+	ask *x509.Certificate
+	ark *x509.Certificate
+}
+
+// SevSnpCerts returns the cached SEV-SNP ASK and ARK certificates.
+func (c *CachedCerts) SevSnpCerts() (ask, ark *x509.Certificate) {
+	return c.ask, c.ark
+}
+
+// createCertChainCache creates a certificate chain cache configmap with the ASK and ARK
+// retrieved from the KDS and returns ASK and ARK.
+// If the configmap already exists, nothing is done and the existing ASK and ARK are returned.
+func (c *Client) createCertChainCache(ctx context.Context, signingType abi.ReportSigner) (ask, ark *x509.Certificate, err error) {
+	c.log.Debugf("Creating certificate chain cache")
+	ask, ark, err = c.getCertChainCache(ctx)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get certificate chain cache: %w", err)
+	}
+	if ask != nil && ark != nil {
+		c.log.Debugf("Certificate chain cache already exists")
+		return ask, ark, nil
+	}
+
+	c.log.Debugf("Retrieving certificate chain from KDS")
+	ask, ark, err = c.kdsClient.CertChain(signingType)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to retrieve certificate chain from KDS: %w", err)
+	}
+
+	askWriter := &strings.Builder{}
+	if err := pem.Encode(askWriter, &pem.Block{Type: "CERTIFICATE", Bytes: ask.Raw}); err != nil {
+		return nil, nil, fmt.Errorf("failed to encode ask: %w", err)
+	}
+	arkWriter := &strings.Builder{}
+	if err := pem.Encode(arkWriter, &pem.Block{Type: "CERTIFICATE", Bytes: ark.Raw}); err != nil {
+		return nil, nil, fmt.Errorf("failed to encode ark: %w", err)
+	}
+
+	c.log.Debugf("Creating certificate chain cache configmap")
+	if err := c.kubeClient.CreateConfigMap(ctx, constants.SevSnpCertCacheConfigMapName, map[string]string{
+		constants.CertCacheAskKey: askWriter.String(),
+		constants.CertCacheArkKey: arkWriter.String(),
+	}); err != nil {
+		return nil, nil, fmt.Errorf("failed to create certificate chain cache configmap: %w", err)
+	}
```

Comment:
```suggestion
	if err := c.kubeClient.CreateConfigMap(ctx, constants.SevSnpCertCacheConfigMapName, map[string]string{
		constants.CertCacheAskKey: askWriter.String(),
		constants.CertCacheArkKey: arkWriter.String(),
	}); err != nil {
		// If ConfiMap already exists, another JoinService instance created the certcache while this operation was running
		// Calling this function again should now load the cached certificates
		if k8serrors.IsAlreadyExists(err) {
			return c.getCertChainCache(ctx context.Context)
		}
		return nil, nil, fmt.Errorf("failed to create certificate chain cache configmap: %w", err)
	}
```
There is a potential data race here if two join-service pods start at the same time while no certificate cache exists.
I doubt this will be a problem in practice when creating new clusters, but could cause minor errors when upgrading from a version that did not implement this caching behavior.
We can prevent this by calling `getCertChainCache` again in case the ConfigMap already exists.